# Part a

## Euler Method

In [7]:
import math
# Initial conditions
x = 0
y = 0

# Step size (h)
h = 0.1

# How many steps to take
steps = 5

print(f"Start: x = {x}, y = {y}, y_exact = {math.tan(x)}, error = {0}")

# Loop to perform Euler's Method
for i in range(steps):
    # 1. Calculate the slope: f(x, y) = y^2 + 1
    slope = y**2 + 1
    
    # 2. Update y: y_new = y_old + h * slope
    y = y + (h * slope)
    
    # 3. Update x: x_new = x_old + h
    x = x + h

    # 4. Exact solution y_exact = tan (x)
    y_exact = math.tan(x)

    # 5. Error (Difference)
    error = abs(y_exact - y)
    
    print(f"Step {i+1}: x = {round(x, 2)}, y = {y:.5f}, y_exact = {y_exact:.5f}, error = {error:.5f}")

Start: x = 0, y = 0, y_exact = 0.0, error = 0
Step 1: x = 0.1, y = 0.10000, y_exact = 0.10033, error = 0.00033
Step 2: x = 0.2, y = 0.20100, y_exact = 0.20271, error = 0.00171
Step 3: x = 0.3, y = 0.30504, y_exact = 0.30934, error = 0.00430
Step 4: x = 0.4, y = 0.41435, y_exact = 0.42279, error = 0.00845
Step 5: x = 0.5, y = 0.53151, y_exact = 0.54630, error = 0.01479


As $x$ increases, the Euler method (being a first-order method) begins to lag behind the actual curve because it assumes the slope is constant across the entire interval $h$.

## Runge-Kutta Method

In [18]:
import math
# Initial conditions
x = 0
y = 0

# Step size (h)
h = 0.1

# How many steps to take
steps = 5

print(f"Start: x = {x}, y = {y}, y_exact = {math.tan(x)}, error = {0}")

# Loop to perform RK2 (Heun's Method)
for i in range(steps):
    # 1. Calculate k1 (Slope at the start)
    # This is exactly the same as the Euler slope: k1 = f(x, y) = y^2 + 1
    k1 = y**2 + 1
    
    # 2. Predict y at the end of the interval (Euler step)
    # We need this temporary y to find the slope at the end
    y_end_predict = y + h * k1
    
    # 3. Calculate k2 (Slope at the end): 
    # We use the x at the end (x + h) and the predicted y: k2 = f(x+h, y_end_predict)
    k2 = y_end_predict**2 + 1
    
    # 4. Update y: Average of start slope (k1) and end slope (k2)
    y = y + (h / 2) * (k1 + k2)
    
    # 5. Update x: x_new = x_old + h
    x = x + h
    
    # 6. Exact solution y_exact = tan(x)
    y_exact = math.tan(x)
    
    # 7. Error (Difference)
    error = abs(y_exact - y)
    
    print(f"Step {i+1}: x = {round(x, 2)}, y = {y:.5f}, y_exact = {y_exact:.5f}, error = {error:.5f}")

Start: x = 0, y = 0, y_exact = 0.0, error = 0
Step 1: x = 0.1, y = 0.10050, y_exact = 0.10033, error = 0.00017
Step 2: x = 0.2, y = 0.20304, y_exact = 0.20271, error = 0.00033
Step 3: x = 0.3, y = 0.30981, y_exact = 0.30934, error = 0.00048
Step 4: x = 0.4, y = 0.42341, y_exact = 0.42279, error = 0.00062
Step 5: x = 0.5, y = 0.54702, y_exact = 0.54630, error = 0.00072


## Comparision

Using the same step size ($h=0.1$) and number of steps (5), the Runge-Kutta (RK2) method is significantly more accurate than the Euler method. At $x=0.5$, the Euler method has an error of approximately 0.015, while the RK2 method has an error of only 0.0007, demonstrating the superior convergence of higher-order methods.

# Part b